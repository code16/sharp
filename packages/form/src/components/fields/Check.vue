<template>
    <div class="SharpCheck">
        <label class="SharpCheck__label">
            <input class="SharpCheck__input" type="checkbox" @change="handleCheck" :checked="value" :disabled="readOnly">
            <span class="SharpCheck__appearance">
                <svg class="SharpCheck__checkmark" width="12" height="9" viewBox="0 0 12 9" fill-rule="evenodd">
                    <path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z"></path>
                </svg>
            </span>
            <span class="SharpCheck__label-text">{{text}}</span>
        </label>
    </div>
</template>

<script>
    export default {
        name: 'SharpCheck',
        props: {
            value:Boolean,
            text:String,
            readOnly: Boolean,
        },
        methods: {
            handleCheck(e) {
                this.$emit('input', e.target.checked);
            }
        }
    }
</script>