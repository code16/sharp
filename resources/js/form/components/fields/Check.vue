<script setup lang="ts">
    import { FormCheckFieldData } from "@/types";
    import { FormFieldEmits, FormFieldProps } from "@/form/types";
    import { Checkbox } from "@/components/ui/checkbox";
    import { Label } from "@/components/ui/label";
    import FormFieldLayout from "@/form/components/FormFieldLayout.vue";

    const props = defineProps<FormFieldProps<FormCheckFieldData>>();
    const emit = defineEmits<FormFieldEmits<FormCheckFieldData>>();
</script>

<template>
    <FormFieldLayout v-bind="props" v-slot="{ id, ariaDescribedBy }">
        <div class="group/control flex items-center gap-x-3 h-10">
            <Checkbox
                :id="id"
                :model-value="!!value"
                :disabled="field.readOnly"
                :aria-describedby="ariaDescribedBy"
                @update:model-value="emit('input', $event as boolean)"
            />
            <Label class="py-1 font-medium" :for="id">
                {{ field.text }}
            </Label>
        </div>
    </FormFieldLayout>
</template>
