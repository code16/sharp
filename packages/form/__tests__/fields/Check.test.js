import Vue from 'vue';
import Check from '../../src/components/fields/Check.vue';


describe('check-field', () => {
    Vue.component('sharp-check', Check);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-check :value="true" :read-only="readOnly" @input="inputEmitted($event)"></sharp-check>
            </div>
        `
    });

    test('can mount Check field', async () => {
        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount "read only" Check field', async () => {
        await createVm({
            propsData: {
                readOnly: true
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('emit event on checkbox changed & correct value', async () => {
        let inputEmitted = jest.fn();

        await createVm({
            methods: {
                inputEmitted
            }
        });

        let checkbox = document.querySelector('input');
        checkbox.dispatchEvent(new Event('change', { bubbles: true }));
        expect(inputEmitted).toHaveBeenCalledTimes(1);
        expect(inputEmitted).toHaveBeenCalledWith(true);
    });
});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],

        props:['readOnly']
    });

    await Vue.nextTick();

    return vm.$children[0];
}