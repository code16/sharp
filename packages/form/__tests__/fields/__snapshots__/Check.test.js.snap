// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check-field can mount "read only" Check field 1`] = `
<div id="app">
	<div class="SharpCheck">
		<label class="SharpCheck__label">
			<input type="checkbox" disabled="disabled" class="SharpCheck__input">
			<span class="SharpCheck__appearance">
				<svg width="12" height="9" viewBox="0 0 12 9" fill-rule="evenodd" class="SharpCheck__checkmark">
					<path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z"></path>
				</svg>
			</span>
			<span class="SharpCheck__label-text"></span>
		</label>
	</div>
</div>
`;

exports[`check-field can mount Check field 1`] = `
<div id="app">
	<div class="SharpCheck">
		<label class="SharpCheck__label">
			<input type="checkbox" class="SharpCheck__input">
			<span class="SharpCheck__appearance">
				<svg width="12" height="9" viewBox="0 0 12 9" fill-rule="evenodd" class="SharpCheck__checkmark">
					<path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z"></path>
				</svg>
			</span>
			<span class="SharpCheck__label-text"></span>
		</label>
	</div>
</div>
`;
